locust/test/test_main.py

```diff
@@ -3,30 +3,128 @@
 import json
 import os
 import platform
+import re
+import select
 import signal
 import socket
 import subprocess
 import sys
 import textwrap
+import time
 import unittest
 from subprocess import DEVNULL, PIPE, STDOUT
 from tempfile import TemporaryDirectory
+from typing import IO, Callable, Optional, cast
 from unittest import TestCase
 
 import gevent
 import psutil
 import requests
 from pyquery import PyQuery as pq
+from requests.exceptions import RequestException
 
 from .mock_locustfile import MOCK_LOCUSTFILE_CONTENT, mock_locustfile
 from .util import get_free_tcp_port, patch_env, temporary_file
 
 SHORT_SLEEP = 2 if sys.platform == "darwin" else 1  # macOS is slow on GH, give it some extra time
 
 
+class PollingTimeoutError(Exception):
+    """Custom exception for polling timeout."""
+
+    pass
+
+
+def poll_until(condition_func: Callable[[], bool], timeout: float = 10, sleep_time: float = 0.01) -> None:
+    """
+    Polls the condition_func at regular intervals until it returns True or timeout is reached.
+    Fixed short sleep time ensures faster polling with early exit when condition is met.
+    """
+    start_time = time.time()
+    while time.time() - start_time < timeout:
+        if condition_func():
+            return
+        time.sleep(sleep_time)
+
+    raise PollingTimeoutError(f"Condition not met within {timeout} seconds")
+
+
 def is_port_in_use(port: int) -> bool:
     with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
-        return s.connect_ex(("localhost", port)) == 0
+        try:
+            s.bind(("localhost", port))
+            return False
+        except OSError:
+            return True
+
+
+def web_interface_ready(host: str, port: int) -> bool:
+    try:
+        response = requests.get(f"http://{host}:{port}/", timeout=0.1)
+        return response.status_code == 200
+    except requests.RequestException:
+        return False
+
+
+def all_users_spawned(proc, output):
+    if platform.system() == "Windows":
+        while True:
+            line = proc.stderr.readline()
+            if not line:
+                break
+            output.append(line)
+            if "All users spawned" in line:
+                return True
+    else:
+        stderr = proc.stderr
+        ready, _, _ = select.select([stderr], [], [], 0.1)
+        if ready:
+            line = stderr.readline()
+            output.append(line)
+            if "All users spawned" in line:
+                return True
+    return False
+
+
+def read_output(proc: subprocess.Popen) -> list[str]:
+    output: list[str] = []
+
+    stdout = cast(IO[str], proc.stdout)
+
+    ready, _, _ = select.select([stdout], [], [], 0.1)
+    if ready:
+        line = stdout.readline().strip()
+        output.append(line)
+
+    return output
+
+
+def read_nonblocking(proc: subprocess.Popen, output: list[str]) -> bool:
+    if platform.system() == "Windows":
+        if proc.stdout is None or proc.stderr is None:
```

Comment:
are you doing this just to keep the linter happy? if so, you can just do `assert proc.stdout` and do it before checking the platform, so you only need to do it once.